parsec/core/fs/remote_loader.py

```diff
@@ -417,20 +425,51 @@ async def load_manifest(
             FSDeviceNotFoundError
             FSInvalidTrustchainError
         """
+        return await self._load_manifest(
+            entry_id,
+            version=version,
+            timestamp=timestamp,
+            expected_backend_timestamp=expected_backend_timestamp,
+        )
+
+    async def _load_manifest(
+        self,
+        entry_id: EntryID,
+        version: Optional[int] = None,
+        timestamp: Optional[DateTime] = None,
+        expected_backend_timestamp: Optional[DateTime] = None,
+        workspace_entry: Optional[WorkspaceEntry] = None,
+    ) -> BaseRemoteManifest:
         if timestamp is not None and version is not None:
             raise FSError(
                 f"Supplied both version {version} and timestamp `{timestamp}` for manifest "
                 f"`{entry_id}`"
             )
         # Download the vlob
-        workspace_entry = self.get_workspace_entry()
+        current_workspace_entry = self.get_workspace_entry()
+        if workspace_entry is None:
+            workspace_entry = self.get_workspace_entry()
         with translate_backend_cmds_errors():
             rep = await self.backend_cmds.vlob_read(
                 workspace_entry.encryption_revision,
                 entry_id,
                 version=version,
                 timestamp=timestamp if version is None else None,
             )
+        # Special case for loading manifest while in maintenance
+        if (
+            rep["status"] == "in_maintenance"
+            and workspace_entry.encryption_revision == current_workspace_entry.encryption_revision
+        ):
+            previous_workspace_entry = await self.get_previous_workspace_entry()
```

Comment:
Shouldn't we provide explicitly the revision number of the workspace entry we want ?
Currently we return the n-1 workspace entry version, which might be `current_workspace_entry` if in the meantime (given we are calling an async function) a reencryption has started !

I'm not sure if this behavior is buggy (it would say we call `_load_manifest` with `current_workspace_entry == previous_entry` so this call try to do the same thing it caller has tried, then call `_load_manifest` another time, this time with a correct `previous_entry` value), but at least it should be documented given how slippy the slope is ^^

Reply:
The alternative would be to pass a `encryption_revision` parameter to `get_previous_workspace_entry`.
At first this seems like a cleaner api, and you have the guaranty you won't be spiraling in nested `_load_manifest` callbacks.
On the other hand the "double reencryption" scenario I was talking to is less transparent (given the second call to `_load_manifest` end up with the backend saying we are requesting a n-2 encryption revision which never allowed)

The scenario:
```
encryption_revision == 1
<---- start encryption, encription_revision == 2
call _load_manifest(previous_encription_revision=None)
    --> get_manifest(encription_revision=2)
   <-- ko, revision not available (should use encription_revision==1)
    call _load_manifest(previous_encription_revision=1)
        <---- start encryption, encription_revision == 3
        --> get_manifest(encription_revision=1)
        <-- ko, revision not available (should use encription_revision==2)
        ???
```
At `???` we could choose to do another `_load_manifest(previous_revision=2)` (the behavior of the current code) or to raise an `FSBadEncryptionRevision` exception (the behavior we would most likely have if `get_previous_workspace_entry` would take a revision number)

Another scenario would be:
```
encryption_revision == 1
<---- start reencryption, encription_revision == 2
call _load_manifest(previous_encription_revision=None)
    --> get_manifest(encription_revision=2)
   <-- ko, revision not available (should use encription_revision==1)
    call _load_manifest(previous_encription_revision=1)
        <---- end reencryption
        --> get_manifest(encription_revision=1)
        <-- ko, revision not available (should use encription_revision==2)
        ???
```
Here the current behavior is to raise a `FSBadEncryptionRevision` (given we no longer are in maintenance), so closer to the behavior with `get_previous_workspace_entry` taking a revision number.

I didn't check how this `FSBadEncryptionRevision` is handled, so no definitive opinion on this though ;-)

Reply:
> Currently we return the n-1 workspace entry version, which might be `current_workspace_entry` if in the meantime (given we are calling an async function) a reencryption has started !

...which might be the PREVIOUS `current_workspace_entry`, but the next call to `get_workspace_entry()` should return a new `current_workspace_entry` which has an encryption revision strictly greater than `previous_workspace_entry`. 

I'll add the following assert to make sure we don't fall in an infinite loop because of some other bug:
```python
assert previous_workspace_entry.encryption_revision < self.get_workspace_entry().encryption_revision
```

Reply:
5769bda